packages/serverpod/lib/src/server/endpoint_dispatch.dart

```diff
@@ -336,13 +359,102 @@ class MethodStreamCallContext {
 
+  OnRevokedAuthenticationCallback? _onRevokedAuthentication;
+
+  MessageCentralListenerCallback? _revokedAuthenticationCallback;
+
+  final Session _session;
+
+  AuthenticationInfo? _authenticationInfo;
+
   /// Creates a new [MethodStreamCallContext].
   MethodStreamCallContext({
     required this.method,
     required this.arguments,
     required this.inputStreams,
     required this.endpoint,
-  });
+    required Session session,
+  }) : _session = session;
+
+  /// Initializes the context.
+  Future<void> init() async {
+    var authenticationIsRequired =
+        endpoint.requireLogin || endpoint.requiredScopes.isNotEmpty;
+
+    if (authenticationIsRequired) {
+      var authenticationInfo = await _session.authenticated;
+      if (authenticationInfo == null) {
+        throw NotAuthorizedException(
+          ResultAuthenticationFailed.unauthenticated(
+            'Authentication was required but no authentication info could be retrieved.',
+          ),
+        );
+      }
+
+      _authenticationInfo = authenticationInfo;
+
+      void localRevokedAuthenticationCallback(event) async {
+        var authenticationRevokedReason = switch (event) {
+          RevokedAuthenticationUser _ =>
+            AuthenticationFailureReason.unauthenticated,
+          RevokedAuthenticationAuthId revokedAuthId =>
+            revokedAuthId.authId == authenticationInfo.authId
+                ? AuthenticationFailureReason.unauthenticated
+                : null,
+          RevokedAuthenticationScope revokedScopes => revokedScopes.scopes.any(
+              (s) => endpoint.requiredScopes.map((s) => s.name).contains(s),
+            )
+                ? AuthenticationFailureReason.insufficientAccess
+                : null,
+          _ => null,
+        };
+
+        if (authenticationRevokedReason != null) {
+          var localOnRevokedAuthentication = _onRevokedAuthentication;
+          if (localOnRevokedAuthentication == null) {
+            throw StateError(
+              'Authentication was revoked but no callback was set.',
+            );
+          }
+
+          localOnRevokedAuthentication(authenticationRevokedReason);
+        }
+      }
+
+      _session.messages.addListener(
+        MessageCentralServerpodChannels.revokedAuthentication(
+          authenticationInfo.userId,
+        ),
+        localRevokedAuthenticationCallback,
+      );
+
+      _revokedAuthenticationCallback = localRevokedAuthenticationCallback;
+    }
+  }
+
+  /// Call when the stream call is cancelled to do proper cleanup.
+  void onStreamCancelled() {
+    var localRevokedAuthenticationCallback = _revokedAuthenticationCallback;
+    var localAuthenticationInfo = _authenticationInfo;
+
+    if (localRevokedAuthenticationCallback != null &&
+        localAuthenticationInfo != null) {
+      _session.messages.removeListener(
+        MessageCentralServerpodChannels.revokedAuthentication(
+          localAuthenticationInfo.userId,
+        ),
```

Comment:
Suggestion: Save this locally and simply remove it if it is set.

Since we are not in a static context we might as well save it.
I'm also not sure we are guaranteed to produce the same reference here?